Config flow for homekit

bdraco opened the pull request:
## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Allows multiple homekit bridges to run

HAP-python state is now stored at `.storage/homekit.{entry_id}.state`
aids is now stored at `.storage/homekit.{entry_id}.aids`

Overcomes 150 device limit by supporting
multiple bridges.

Name and port are now automatically allocated
to avoid conflicts which was one of the main
reasons pairing failed.

YAML configuration remains available in order to offer entity
specific configuration.

Entries created by config flow can add and remove
included domains and entities without having to restart

<img width="855" alt="Screen Shot 2020-04-22 at 5 13 06 PM" src="https://user-images.githubusercontent.com/663432/80039067-906cf800-84bc-11ea-89fa-6bb124a48b84.png">


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml

```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/13140

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

bdraco commented:
@Jc2k  I'm creating a device registry entry for the virtual bridge.  This should give you enough information to exclude it from `homekit_controller` discovery

Jc2k commented:
This is super cool!

I don't have anywhere near this many devices but I will be switching to this so there is one less thing to have in my configuration.yml! :-)

I'll have a think about how best make it work in homekit_controller, if it doesn't already. I think we block on the [model](https://github.com/home-assistant/core/blob/dev/homeassistant/components/homekit_controller/config_flow.py#L199) right now? (It was like that when I got here and I've never had need to change it). Will that still work or does the model change per bridge?

Does this also mean that we can automatically export entities to HomeKit without a restart when adding a new integration, and automatically remove them if an integration is removed? Or is that super exciting future on the horizon?

bdraco commented:
> This is super cool!
> 
> I don't have anywhere near this many devices but I will be switching to this so there is one less thing to have in my configuration.yml! :-)
> 
> I'll have a think about how best make it work in homekit_controller, if it doesn't already. I think we block on the [model](https://github.com/home-assistant/core/blob/dev/homeassistant/components/homekit_controller/config_flow.py#L199) right now? (It was like that when I got here and I've never had need to change it). Will that still work or does the model change per bridge?
I didn't even realize that code was in there as it never worked for me. ... Did some digging and the model that is actually exported is:

`const.py:BRIDGE_MODEL = "Bridge"`

It has been like that for a long time.  Before #14114 it was

`BRIDGE_MODEL = 'homekit.bridge'`

Ideally we can now look in the device registry for the MAC address of the bridge and exclude it after this goes in.


> 
> Does this also mean that we can automatically export entities to HomeKit without a restart when adding a new integration, and automatically remove them if an integration is removed? Or is that super exciting future on the horizon?

Yes, you can add and remove domains and exclude entities in the options flow

Jc2k commented:
Ah, i never noticed because my bridge is always paired and that code was always there so i never checked! D'oh. I guess it was for the old Homebridge? Yes, the new way sounds much better.

So say i had already exported the light domain, and then a day later i paired a new light builb. Would it would show up immediately wihtout restarting?

bdraco commented:
> Ah, i never noticed because my bridge is always paired and that code was always there so i never checked! D'oh. I guess it was for the old Homebridge? Yes, the new way sounds much better.

I think that is spot on
> 
> So say i had already exported the light domain, and then a day later i paired a new light builb. Would it would show up immediately wihtout restarting?

You would need to go into the options flow and make sure the 'light' domain was included and then it would show up.  Running the options flow forces a reload of the HAP server.  I could probably add a service to do it as well but its easy enough to just run the options flow.

Jc2k commented:
I think there is a service that does it already, but its not as fancy as having it hooked new entity registry entries 😁

joogps commented:
Wouldn't it be able to define `entity_config` for the other bridges? I made a pull request that expands the remote functionality (#34570), but Apple limits HomeKit remote to one TV per bridge. Multiple bridges would be the solution, but `entity_config` would need to be available for configuration. Maybe you could redirect manual configuration to a config flow? I've done that with the `panasonic_viera` integration and it works flawlessly.

Jc2k commented:
@joogps What if you used the new [triggers API](https://developers.home-assistant.io/docs/device_automation_trigger/) so that manual configuration yaml wasn't required? I think this is the preferred way to allow users to configure how to handle this sort of thing now isn't it?

With this PR We now create Device Registry entries for the bridges, so you could associate the triggers with that Device Registry entry maybe.

joogps commented:
I didn't think about that. Cool, it might work!

Would the user then need to create automations to associated the keys to the service?

joogps commented:
I'll make a question in the Discord server and hear what they think about it :D

bdraco commented:
We could implement entity config options in multiple config flow paths but that would need to be done in a follow up as this PR is already larger than I had hoped

joogps commented:
Yep, that's what I think I'll do. I'll close mine and see what happens next before creating a new one.

bdraco commented:
Rebase to fix conflict with #34587

joogps commented on homeassistant/components/homekit/__init__.py:
```diff
@@ -134,28 +120,87 @@
+            hass.config_entries.async_update_entry(entry, data=data, options=options)
+            return True
+
+    hass.async_create_task(
```
Here, I think you can make multiple configurations through YAML be taken into consideration. Like this:

```python
for conf in config[DOMAIN]:
        hass.async_create_task(
            hass.config_entries.flow.async_init(
                DOMAIN, context={"source": SOURCE_IMPORT}, data=conf
            )
        )
```

Of course, you'd also need to change the config schema to allow this.

bdraco replied:
@joogps I didn't do this in order to avoid making this a breaking change.

joogps replied:
Ohh, ok. Didn't think about that 😅

joogps replied:
But would it affect users? I think (and I'm probably wrong) that single configurations (without dashes) would still work fine.

bdraco replied:
We would need to migrate the aid storage and the homekit state files as well. Could do this a few versions from now once everything has already been imported so only people who didn't upgrade though the versions would loose their custom names and room assignments.

bdraco replied:
I'll see if I can come up with a way to support multiple yaml entries without making this a breaking change.  Unfortunately that will make this PR quite a bit larger

joogps commented:
Great! Thanks, @bdraco! 🥰

bdraco commented:
This was passing before the CI broke
```
----------- coverage: platform linux, python 3.8.1-final-0 -----------
Name                                                        Stmts   Miss  Cover   Missing
-----------------------------------------------------------------------------------------
homeassistant/components/homekit/__init__.py                  264     30    89%   248-249, 256-258, 264-282, 287, 317, 320-324, 360, 437-440, 505, 609-619
homeassistant/components/homekit/accessories.py               222      2    99%   294, 366
homeassistant/components/homekit/aidmanager.py                 76      4    95%   65-66, 130, 137
homeassistant/components/homekit/config_flow.py               110      2    98%   156-157
homeassistant/components/homekit/const.py                     174      0   100%
homeassistant/components/homekit/type_covers.py               155      0   100%
homeassistant/components/homekit/type_fans.py                 100      0   100%
homeassistant/components/homekit/type_lights.py                91      0   100%
homeassistant/components/homekit/type_locks.py                 40      0   100%
homeassistant/components/homekit/type_media_players.py        189      0   100%
homeassistant/components/homekit/type_security_systems.py      37      0   100%
homeassistant/components/homekit/type_sensors.py              129      0   100%
homeassistant/components/homekit/type_switches.py             109      0   100%
homeassistant/components/homekit/type_thermostats.py          230      1    99%   563
homeassistant/components/homekit/util.py                      165     12    93%   320-325, 341-342, 354-357
-----------------------------------------------------------------------------------------
TOTAL                                                        2091     51    98%


Results (19.48s):
     166 passed
```

bdraco committed:
Config flow for homekit

Allows multiple homekit bridges to run

HAP-python state is now stored at .storage/homekit.{entry_id}.state
aids is now stored at .storage/homekit.{entry_id}.aids

Overcomes 150 device limit by supporting
multiple bridges.

Name and port are now automatically allocated
to avoid conflicts which was one of the main
reasons pairing failed.

YAML configuration remains available in order to offer entity
specific configuration.

Entries created by config flow can add and remove
included domains and entities without having to restart

bdraco commented:
I'm going to do a soft reset and squash since this conflicts are becoming unmanageable.

balloob commented on homeassistant/components/homekit/__init__.py:
```diff
@@ -166,20 +231,100 @@ async def async_setup(hass, config):
+                    "HomeKit is not running. Either it is waiting to be "
+                    "started or has been stopped."
+                )
+                return
```
This should be a continue I think ?

bdraco replied:
yes, fixed

balloob commented on homeassistant/components/homekit/__init__.py:
```diff
@@ -50,110 +46,179 @@
+
+    if CONF_ENTRY_INDEX in conf and conf[CONF_ENTRY_INDEX] == 0:
+        _LOGGER.debug("Migrating legacy HomeKit data for %s", name)
+        migrate_filesystem_state_data_for_primary_imported_entry_id(
```
Run in executor

bdraco replied:
done

balloob commented on homeassistant/components/homekit/__init__.py:
```diff
@@ -480,10 +543,14 @@ def _start(self, bridged_states):
 
         if not self.driver.state.paired:
             show_setup_message(
```
Why is this not part of the config flow? So we finish the flow when the pairing is done. We can do this with external step. I think that we might have to adjust external step to also work without a url. 

This PR is already big, so we can also postpone this for a future PR.

bdraco replied:
Pairing can happen outside of the config flow.  If they unpair they can repair on another device.

balloob replied:
Couldn't we start a new pair config flow in that case if we get unpaired that just updates the existing flow?

bdraco replied:
I've been watching https://github.com/home-assistant/architecture/issues/377 to see how it pans out as it might be a similar use case.

I'll do a followup pending the outcome of the above (it could turn out to not be relevant to this one).

balloob commented on homeassistant/components/homekit/config_flow.py:
```diff
@@ -0,0 +1,288 @@
+        default_domains = [] if self._async_current_entries() else DEFAULT_DOMAINS
+        setup_schema = vol.Schema(
+            {
+                vol.Optional(CONF_AUTO_START, default=DEFAULT_AUTO_START): bool,
```
Can we only add this option if `self.show_advanced_options` ? With the `started` event combined with entity registry restoring unavailable entities, it's not necessary anymore

bdraco replied:
Done, looks the the frontend (https://github.com/home-assistant/frontend/pull/5612) isn't merged yet so I wrote coverage for the basic path

balloob replied:
Yeah Bram was waiting until we didn't need frontend releases done for 109 anymore.

balloob commented on homeassistant/components/homekit/strings.json:
```diff
@@ -0,0 +1,54 @@
+{
+   "title" : "HomeKit Bridge",
```
Let's drop this so it will default to the manifest name (HomeKit)

bdraco replied:
Removed the title.  and changed `manifest` to add `Bridge`.  I'm hoping this clears up the al the "you really want HomeKit Controller" confusion.

balloob replied:
Oh well in that case we might need to put it back as title, because we need to translate the word bridge.

balloob commented on homeassistant/components/homekit/strings.json:
```diff
@@ -0,0 +1,54 @@
+         },
+         "advanced" : {
+            "data" : {
+               "auto_start" : "Autostart (disable if using Z-Wave or other delayed start system)",
```
Use as much references as possible.

```suggestion
               "auto_start" : "[%key:component::homekit::config::step::user::data::auto_start%]",
```

bdraco replied:
Nice, didn't know that worked :)

balloob replied:
I'm working on popularizing it. Have a draft pr to migrate bunch of config flow translations. I just have been getting myself into a lot of never ending projects lately, so will get to it soon 🙃

bdraco committed:
Fix services as there are multiple now

bdraco committed:
migrate in executor

bdraco committed:
drop title from strings

bdraco committed:
Update homeassistant/components/homekit/strings.json

Co-authored-by: Paulus Schoutsen <paulus@home-assistant.io>

bdraco committed:
Merge branch 'homekit_config_entry' of github.com:bdraco/home-assistant into homekit_config_entry

bdraco committed:
Make auto_start advanced mode only, add coverage

bdraco commented:
@balloob Thanks for the review.

bdraco committed:
put back title

balloob commented on homeassistant/components/homekit/translations/en.json:
```diff
@@ -21,7 +22,7 @@
 			 "data" : {
-				"auto_start" : "Autostart (disable if using Z-Wave or other delayed start system)",
+				"auto_start" : "[%key:component::homekit::config::step::user::data::auto_start%]",
 				"safe_mode" : "Safe Mode (enable only if pairing fails)",
```
Don't forget to use references for the other ones

bdraco replied:
found another one :)

bdraco committed:
more references

bdraco committed:
delete port since manual config is no longer needed

bdraco commented:
Test failure is unrelated (its in august)

Fixing here https://github.com/home-assistant/core/pull/34998

MartinHjelmare commented on homeassistant/components/homekit/strings.json:
```diff
@@ -0,0 +1,54 @@
+{
+	"title" : "HomeKit Bridge",
```
We should avoid 8 space indentation. It pushes the content too far to the right.

bdraco replied:
I sent them though `python -m json.tool`

https://github.com/home-assistant/core/pull/35056
